_posts/2023-02-03_medical-imaging-with-nix/medical-imaging-with-nix.Rmd

```diff
@@ -0,0 +1,211 @@
+
+So this points to the biggest advantage of conda over nix, when conda does not have a package it gets it from pypi, nixpkgs cannot fail over
+to grabbing from pypi and installing with pip, it also can't do dependency solving, if the nixpkgs version of a python lib isn't compatible with
+another, you have to go find a satisfactory version yourself (or cheat and [lie about the version requirements](https://github.com/LKS-CHART/medical-imaging-nix/blob/2e811b3eb9b931f708d5f4b3658506b496ce24a0/container/overlay.nix#L71)^
+[I was reminded after the first draft of this post about the existence of `pythonRelaxDeps`, a way to automatically soften version requirements,
+I may switch to use this in the future.]). This is painful. There are two nix projects that have aimed to address this
+problem, [mach-nix](https://github.com/DavHau/mach-nix) which has been abandoned^[although hopefully returning as part of https://github.com/nix-community/dream2nix in the future], and [poetry2nix](https://github.com/nix-community/poetry2nix) which may solve
+some of my woes but I haven't tried it yet.
+
+Another advantage alluded to already is speed. Compiling things takes time, so unless you're always getting prebuilt binaries from the
+nix servers you can be in for long build times. Nix is good at not duplicating build work, but sometimes rebuilding is unavoidable. Say
+you want to use a newer version of cuda, or gcc, most of your environment will need to be rebuilt^[ a colleague suggests that we create a build
+farm where a background process builds new versions of things so that there are prebuilt versions always available, but that is an infrastructure
+investment that hasn't felt worth it yet. Apparently the issue is that CUDA is technically not free software and by policy the nix
+build farms won't build it for us. ]
+
+Finally, the last difficulty is with irreducible system dependencies. Nixpkgs, when you're not using the nix operating system, does
+not interact well with graphics drivers, there is a wrapper project I use called [NixGL](https://github.com/guibou/nixGL) which gives 
+you access to graphics drivers and allows you to run programs that use CUDA^[you need to start programs with `nixGL<driver> <program>` which
+is irritating, but I think can be fixed in my flake]. However this has meant I needed to hardcode my graphics
+driver into the nix flake, severely hindering portability^[Ok you'd need to edit one line in the flake but that's enough of a barrier
+to discourage users. If anyone in the nix community can help me solve this problem I'd be extremely grateful]. 
+
+## Is this worth it?
+
+I don't know. I've sunk considerable cost into building this environment so one might reasonably expect me to be quite biased. 
+
+Would cutting down my effort by 75% at the risk of some amount of computational irreproducibility and unportability be worth it?
+
+Is this a case of good enough practices trumping best practices? 
```

Comment:
I feel like we read a paper/blog about "good enough" vs "best" practices at some point for journal club @derekbeaton 
Maybe we could add a link to it here?

Reply:
I was going to put that link in and forgot, https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005510, is this the one you all read?

Reply:
Yes, it was this one!!

Reply:
added the link, great spot and suggestion